### PR TITLE
Introduce read-only wallet for xpubs.

### LINKED ISSUE #489: Allow ASB to move money to a zpub instead of the internal wallet

### PR DESCRIPTION
**Pull request by bonomat**
At the end of a swaps (may it be redeemed, refunded, or punished), the Bitcoin funds should be transferred to an address controlled by the according party.
This wallet should be decoupled from the protocol wallet, i.e. the wallet which will be used to sign internal transactions.
In this PR, we introduce a `personal_wallet` which is meant for exactly this.

Resolves #489 

Some notes: 

1. `zpub-support`: a `zpub` is an `xpub` which implies native segwit support, i.e. an address derived from a `zpub` key should be prefixed with `bc1` (if mainnet). 
Unfortunately [bdk](https://github.com/bitcoindevkit/bdk) does not support `zpub` keys. However, we can achieve the same with `xpub` keys as well. By implying Bip84 descriptors, any key derived from this address will be a `bech32` (native segwit) address. 
3. optional `xpub`-entry for Asb config: A user running the ASB can optionally provide an `xpub` key. We will derive `redeem` and `punish` addresses from this. Since `xpub` keys are read-only, this has the implication that the `withdraw-all` command is useless. If, however, the user does not provide an `xpub` key, we use our internal seed to derive a wallet. In this case, the `withdraw-all` command is still useful :) 
4. `xpub`- support for Swap Cli it wasn't part of the ticket so I only pulled it through to the binary but do not expose it yet to the user. I guess this will need some more thinking, e.g. does an xpub make sense for the Swap Cli or would a single refund address be sufficient? A single address has the disadvantage of making our tests more complicated, i.e. if we have a full wallet, we can just _connect_ it to the blockchain and query for the balance while with an address we would need to build a bit more logic for this. I propose to create a follow-up ticket on this.

### CODE REVIEW

File: swap/src/protocol/alice/event_loop.rs [hunk 1]
```diff
@@ -153,14 +156,14 @@ where
...
                                 .estimate_fee(bitcoin::TxRedeem::weight(), btc)
                                 .await;
-                            let tx_punish_fee = self.bitcoin_wallet
+                            let tx_punish_fee = self.internal_bitcoin_wallet
                                 .estimate_fee(bitcoin::TxPunish::weight(), btc)
                                 .await;
-                            let redeem_address = self.bitcoin_wallet.new_address().await;
-                            let punish_address = self.bitcoin_wallet.new_address().await;
+                            let redeem_address = self.personal_bitcoin_wallet.new_address().await;
+                            let punish_address = self.personal_bitcoin_wallet.new_address().await;
```
Comment (thomaseizinger) [inline swap/src/protocol/alice/event_loop.rs:166]: Passing around two instead of one wallet is quite the invasive change. What if we instead introduce functions such as `new_redeem_address` and `new_punish_address` and decide internally within `bitcoin::Wallet`, where we derive these addresses from? I think that would be less invasive because it allows us to handle things within `bitcoin::Wallet`.

File: swap/src/bin/swap.rs [hunk 2]
```diff
@@ -296,18 +298,18 @@ async fn init_bitcoin_wallet(
     .context("Failed to initialize Bitcoin wallet")?;
 
     let wallet_dir = data_dir.join("personal_wallet");
-    let _personal_wallet = bitcoin::wallet::Wallet::new_readonly(
+    let personal_wallet = bitcoin::Wallet::new_readonly(
```
Comment (thomaseizinger) [inline swap/src/bin/swap.rs:301]: Having a "personal wallet" inside the CLI code that is not actually exposed to the user feels like unnecessary complexity to me. I don't want to need to read code that looks like it matter for the user but actually doesn't.

File: swap/src/seed.rs [hunk 3]
```diff
@@ -41,6 +43,19 @@ impl Seed {
...
+    pub fn derive_extended_public_key(&self, network: bitcoin::Network) -> Result<ExtendedPubKey> {
+        let priv_key = self.derive_extended_private_key(network)?;
+        let secp = Secp256k1::new();
+
+        // we take the first child private key to derive a non_hardened xpub key
+        let non_hardened_xprv = priv_key.ckd_priv(&secp, ChildNumber::from_normal_idx(0)?)?;
+
+        let public_key = ExtendedPubKey::from_private(&secp, &non_hardened_xprv);
+        Ok(public_key)
+    }
```
Comment (thomaseizinger) [inline swap/src/seed.rs:57]: So we are able to spend coins sent to this address because it is a child-key of the xprv that we use for the other wallet?

This is not obvious from my point of view. Can we just use the internal wallet for everything if the user didn't specify an xpub instead of deriving one that we happen to know the xprv for?

I am worried that the invariants that need to be maintained here for things to work are too complicated and we end up deriving addresses that we don't actually have the keys for.

File: swap/tests/harness/mod.rs [hunk 4]
```diff
@@ -734,28 +743,15 @@ impl TestContext {
...
-        } else {
-            bail!("Bob in not in xmr redeemed state: {:?}", state);
-        };
-
-        let lock_tx_bitcoin_fee = self
-            .bob_internal_bitcoin_wallet
-            .transaction_fee(lock_tx_id)
-            .await?;
-
-        Ok(self.bob_starting_balances.btc - self.btc_amount - lock_tx_bitcoin_fee)
```
Comment (thomaseizinger) [inline swap/tests/harness/mod.rs]: If you are deleting code like this, can we

a) have a dedicated commit for it
b) explain why we can do it without causing problems?

File: swap/tests/harness/mod.rs [hunk 5]
```diff
@@ -675,36 +683,15 @@ impl TestContext {
...
-        let refund_fee = self
-            .alice_internal_bitcoin_wallet
-            .estimate_fee(TxRefund::weight(), self.btc_amount)
-            .await
-            .expect("To estimate fee correctly");
-
-        let bob_cancelled_and_refunded = btc_balance_after_swap
-            == self.bob_starting_balances.btc - lock_tx_bitcoin_fee - cancel_fee - refund_fee;
-
-        assert!(bob_cancelled_and_refunded);
```
Comment (thomaseizinger) [inline swap/tests/harness/mod.rs]: Same thing here, please explain why it is ok to delete this. Ideally in a dedicated refactoring commit that doesn't involve other functionality.

File: swap/src/asb/config.rs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 use crate::fs::{ensure_directory_exists, system_config_dir, system_data_dir};
 use crate::tor::{DEFAULT_CONTROL_PORT, DEFAULT_SOCKS5_PORT};
 use anyhow::{bail, Context, Result};
+use bdk::bitcoin::util::bip32::ExtendedPubKey;
```
Comment (thomaseizinger) [inline swap/src/asb/config.rs:4]: We also depend on `bitcoin`, no need to import things through the re-export. Please prefer a direct import from the `bitcoin` crate unless you want to be uniform across the codebase and import via bdk everywhere. Then we could remove the `bitcoin` dependency but I think we will need it anyway to activate certain features.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thank you for taking this on so quickly!

I'll structure this message in two parts. First, some general feedback on the implementation and second, my personal opinion on the notes that you provided.

1. Your commits aren't particularly atomic.
- There are TODOs and hardcoded constants which are "resolved" in later commits.
- Refactorings and new features often happen in the same commit.
- Things get changed or renamed in later commits after they have been introduced earlier.
- The message often doesn't explain _why_ something happens. I've noted some of these instances inline.
- Personally, I think a PR is easiest to review if the initial patches are pure refactorings that are topped of with the feature we want. The reason being that I can adjust my mental model of the codebase without having to think about change in functionality. I also don't have to focus as much on correctness if I know it is a pure refactoring because the tests will (hopefully) catch a bug there. Finally, all that is "new" is introduced a single unit that just happens to fit in nicely because you've paved the way with your refactorings.

2. It seems to me that passing around a complete separate instance of `bitcoin::Wallet` is very invasive if all we want to do is derive two addresses from a different xpub. Not sure how it pans out but keeping the difference of where we derive the address from _internal_ to `bitcoin::Wallet` might make all of this a lot easier? It would imply making the `bitcoin::Wallet` aware of protocol terms like `refund` and `redeem` but I don't see why it couldn't do that. It is after all a wallet for a swap application. Anything more general should be upstreamed to bdk. Happy to roll with completely different designs as well but the current one is a fairly bad trade-off in terms of complexity vs. feature gain.

3. Like I commented inline, introducing the concept of a personal wallet to the CLI without exposing it to the user increases complexity for no particular gain. Would be nice if we can minimize the code changes to the CLI here unless we actually plan to expose this feature.

Regarding your notes:

> optional xpub-entry for Asb config: A user running the ASB can optionally provide an xpub key. We will derive redeem and punish addresses from this. Since xpub keys are read-only, this has the implication that the withdraw-all command is useless. If, however, the user does not provide an xpub key, we use our internal seed to derive a wallet. In this case, the withdraw-all command is still useful :)

If `withdraw-all` is pointless with an `xpub` set, I think we should tell the user explicitly when they invoke the command.

> xpub- support for Swap Cli it wasn't part of the ticket so I only pulled it through to the binary but do not expose it yet to the user. I guess this will need some more thinking, e.g. does an xpub make sense for the Swap Cli or would a single refund address be sufficient? A single address has the disadvantage of making our tests more complicated, i.e. if we have a full wallet, we can just connect it to the blockchain and query for the balance while with an address we would need to build a bit more logic for this. I propose to create a follow-up ticket on this.

Agreed that we should be dealing with this feature for the CLI in a separate ticket.

File: swap/src/protocol/alice/event_loop.rs (hunk 1 above)
Comment (bonomat) [reply swap/src/protocol/alice/event_loop.rs:166]: I thought about something similar but decided against it. 
Here are my thoughts: 

1. I thought about adding a new field to `bitcoin::Wallet` i.e. `personal_wallet` so that we can achieve what you proposed. I decided against it because it would destroy the meaning of `bitcoin::Wallet`, because it's not a single `Wallet` anymore but something like `Wallets`... the abstraction did not add-up for me because how do you know from the outside what do you call `balance()`/`sign`/`new_address` on without knowing the internal structure in advance? 
2. I then wanted to have a completely separate wallet implementation for the read-only wallet, i.e. independent from `bitcoin::wallet`. For example, we could have `{crate}::bitcoin::wallets::internal` and `{crate}::bitcoin::wallets::personal`. `personal` could than be an offline-only wallet allowing easier testing key and address derivations. I then ran into problems because I had to assert the balances in the tests. It would also add some duplicated code.
3. I decided to re-use the same wallet which I think turned out quite nice. They are too similar to separate and it allowed me to not have to implement a new wallet for e2e tests which would allow me to assert on the balances. 

Imho this way is best trade-off between asking developers to go yet one level further down.

File: swap/src/seed.rs (hunk 3 above)
Comment (bonomat) [reply swap/src/seed.rs:57]: Yes, we could also just pass around the _same_ wallet twice, e.g. either clone or instantiate it again from `bitcoin::Wallet`. Imho this is harder to understand for a developer than having two keys, one for each wallet. 

> I am worried that the invariants that need to be maintained here for things to work are too complicated and we end up deriving addresses that we don't actually have the keys for.

I don't follow your concern: if we derive an xpub from the user's seed we should always be able to also derive the private key from it. 
If I misunderstood you, can you rephrase please?

File: swap/tests/harness/mod.rs (hunk 4 above)
Comment (bonomat) [reply swap/tests/harness/mod.rs]: > a) have a dedicated commit for it

Not really, it depends on the other change in this commit. 

> b) explain why we can do it without causing problems?

The reason is that the balance of Bob's `personal` bitcoin wallet needs to be 0 after the bitcoin was redeemed. Can add this reasoning into the commit message

File: swap/src/bitcoin/wallet.rs [hunk 7]
```diff
@@ -268,6 +315,7 @@ where
         Ok(Amount::from_sat(balance))
     }
 
+    // TODO(philipp) remove me
```
Comment (bonomat) [inline swap/src/bitcoin/wallet.rs:318]: This todo is wrong 😅

**Comment by bonomat** [comment]
Converted to draft and handed over to @da-kami :) .

File: CHANGELOG.md [hunk 8]
```diff
@@ -28,6 +28,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - A minimum accepted Bitcoin amount for the ASB similar to the maximum amount already present.
   For the CLI the minimum amount is enforced by waiting until at least the minimum is available as max-giveable amount.
 - Added a new argument to ASB: `--json` or `-j`. If set, log messages will be printed in JSON format.
+- New read-only wallet for ASB which enables users to provide a xpub key to derive redeem/punish address from.
+  If not provided in the config, an internal wallet will be used. Users can still withdraw all their funds from the
+  internal wall.
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply CHANGELOG.md:33]: I highly doubt that :)

a) we don't use `cargo tomlfmt` in this repo
b) it doesn't touch markdown files

File: swap/src/bitcoin/wallet.rs [hunk 9]
```diff
@@ -41,21 +43,66 @@ pub struct Wallet<B = ElectrumBlockchain, D = bdk::sled::Tree, C = Client> {
 }
 
 impl Wallet {
-    pub async fn new(
+    pub fn new(
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply swap/src/bitcoin/wallet.rs:46]: Seems to be in the works: https://github.com/rust-lang/rust-clippy/issues/7176

File: CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (bonomat) [reply CHANGELOG.md:33]: Then I'll take everything back and claim the opposite. There is also something wrong with the sentence so I blame the internet monster.

File: swap/src/protocol/alice/event_loop.rs (hunk 1 above)
Comment (thomaseizinger) [reply swap/src/protocol/alice/event_loop.rs:166]: > I thought about adding a new field to bitcoin::Wallet i.e. personal_wallet so that we can achieve what you proposed. I decided against it because it would destroy the meaning of bitcoin::Wallet, because it's not a single Wallet anymore but something like Wallets... the abstraction did not add-up for me because how do you know from the outside what do you call balance()/sign/new_address on without knowing the internal structure in advance?

Do you have to know that? I don't see why you would.

Unless I am missing something, we would only need to remove `new_address` and add new APIs that qualify, what the semantic of this address is:

- `new_redeem_address`
- `new_refund_address`

Like I pointed out in the overall feedback comment above, our `bitcoin::Wallet` is not a general-purpose Bitcoin wallet (unlike `bdk::Wallet`). It is more the wallet component for a swap application. Anything more generic can still live in there obviously but doesn't need to hold us back in creating an abstraction layer that works _better_ for our swap protocol with better being defined as: the protocol doesn't need to care about application semantics like the xpub feature.

We don't have to treat `bitcoin::Wallet` as a single wallet, that is what an instance of `bdk::Wallet` is for. Instead, imagine `bitcoin::Wallet` being a trait and serving as the interface for our protocol to access wallet functionality. What happens under the hood is irrelevant to the protocol as long as the functionality offered is internally consistent, f.e. `balance` doesn't return coins that you can't actually spend.

In the current design, we introduce a lot of complexity by "leaking" into the protocol that some clients (only ASB for now) might want a different wallet for redeem/refund/punish.

File: swap/src/seed.rs (hunk 3 above)
Comment (thomaseizinger) [reply swap/src/seed.rs:57]: > I don't follow your concern: if we derive an xpub from the user's seed we should always be able to also derive the private key from it.

Yes, as long as the rest of the application follows along here.
What I am saying is, for this to work, several parts of the code need to be kept semantically in sync and that part is not obvious!

To demonstrate what I mean: The "internal" wallet will find coins on the chain that it never handed out an address for. It just happened that a different piece of code, outside the wallet, derived an xpub that belongs to the same seed. Hence, the "ownership" of the xpriv within the wallet is broken.

Having to keep this in mind when modifying the code creates unnecessary complexity.

> Imho this is harder to understand for a developer than having two keys, one for each wallet.

I don't follow how this is harder to understand?
If you trace the code and you see the following:

```rust
let internal_wallet = // ...;

let personal_wallet = if xpub { new_personal_wallet(xpub) } else { personal_wallet.clone() };
```

All you need to keep in your head is:

"So if no xpub is set, I can assume the two wallets are the same."

No Bitcoin specific knowledge required about how derivation paths work etc.
You can simply discard the additional information about two wallets existing and conceptually "merge" them in your head.

I'll claim that it is easier to understand what an application does if you can follow the dataflow of its variables than if you need to have domain-specific knowledge to work out that addresses handed out be one wallet will end up in the balance of the other one. That is just counter-intuitive.

File: swap/tests/harness/mod.rs (hunk 4 above)
Comment (thomaseizinger) [reply swap/tests/harness/mod.rs]: > Not really, it depends on the other change in this commit.

It should be part of that commit then.

File: swap/src/bitcoin/wallet.rs [hunk 10]
```diff
@@ -41,21 +43,66 @@ pub struct Wallet<B = ElectrumBlockchain, D = bdk::sled::Tree, C = Client> {
...
+
+    pub fn new_readonly(
+        electrum_rpc_url: Url,
+        wallet_dir: &Path,
+        key: impl DerivableKey<Segwitv0> + Clone,
+        env_config: env::Config,
+    ) -> Result<Self> {
+        // TODO (phh): figure out what exactly this should be. It will influence the
+        // addresses derived from this key
+        let fingerprint = bitcoin::util::bip32::Fingerprint::from_str("2b4497f1")?;
```
[the remark this answers is not in the document]
Comment (thomaseizinger) [reply swap/src/bitcoin/wallet.rs:73]: This definitely needs to be resolved before we can merge this.

File: swap/src/bin/asb.rs [hunk 11]
```diff
@@ -272,27 +281,39 @@ async fn main() -> Result<()> {
...
+        config.bitcoin.electrum_rpc_url.clone(),
+        &wallet_dir,
+        xpub,
+        env_config,
+    )
+    .context("Failed to initialize personal Bitcoin wallet")?;
 
-    wallet.sync().await?;
+    internal_wallet.sync().await?;
+    personal_wallet.sync().await?;
```
Comment (thomaseizinger) [inline swap/src/bin/asb.rs:314]: The user will have to tell us the re-scan height of the key, otherwise we would assume the key hasn't been used before and we will re-use addresses.

File: swap/tests/harness/mod.rs (hunk 4 above)
Comment (bonomat) [reply swap/tests/harness/mod.rs]: >(Do the e2e tests even pass with the previous one but without this one?) 

Yes, because we were asserting something else: that the balance of the other wallet is `starting_balance - btc_amount - fee`.

File: swap/src/bin/asb.rs (hunk 11 above)
Comment (bonomat) [reply swap/src/bin/asb.rs:314]: I disagree with your conclusion: 

If a user provides an xpub key he really should not use this xpub otherwise to get addresses. 
Otherwise we cannot guarantee that we won't re-use addresses. Even if we ask the user for a height. 

Also, this function does not care about the height. 
Does BDK support the feature of only scanning from a particular height?

File: swap/src/seed.rs (hunk 3 above)
Comment (bonomat) [reply swap/src/seed.rs:57]: > What I am saying is, for this to work, several parts of the code need to be kept semantically in sync and that part is not obvious!

There is the seed generation and the wallet instantiation. I don't see other moving parts which need to keep in sync. 

> "So if no xpub is set, I can assume the two wallets are the same."

I might have misunderstood your initial comment when I answered with : 
> Yes, we could also just pass around the same wallet twice, e.g. either clone or instantiate it again from bitcoin::Wallet. Imho this is harder to understand for a developer than having two keys, one for each wallet.

What I wanted to say with this that I find it harder to understand what you pointed out (We know we have different understandings about code complexity so I won't argue about this any further. :) 
):  

> "So if no xpub is set, I can assume the two wallets are the same."

I'm coming to a different conclusion: the code here is wrong. 
The two wallets should not be the same. 
We should not re-use the same xprv-key to derive the xpub for the personal wallet as there should not be a relation between them. We should just derive a different xprv-key from the seed for the personal wallet.

Comment (thomaseizinger) [reply swap/src/seed.rs:57]: >  The two wallets should not be the same.

But at the moment (without this feature) they are, right? Is your claim that the current code and behaviour is wrong?

My goal would have been: don't "fix" something that is not broken. If we simply clone the original wallet, the xpub functionality is layered on top of what we currently have, without any changes in functionality if it is disabled. Given that we seem to be happy with how the code currently works, I don't see a reason to not do that.

> We should not re-use the same xprv-key to derive the xpub for the personal wallet as there should not be a relation between them. We should just derive a different xprv-key from the seed for the personal wallet.

That would address my concerns about "ownership" of the keys with each wallet. However, I think we care getting off-track here with the feature design in terms of what the original motivation is. I am not sure the concept of a personal vs internal wallet overall makes a whole lot of sense.

1. At the moment, the ASB doesn't need a Bitcoin wallet for anything other than receiving money (the other protocol is still in the works). Hence, even if you want to keep two wallets, there shouldn't be a need to pass a personal-wallet into any Alice / ASB related code.
2. Once we allow trading in both directions, Alice needs to be able to transfer BTC coins into a protocol as well. For that, it will need an xprv to sign transactions (i.e. an xpub is not enough).

If we have a personal and an internal wallet and the xpub feature is not used, which one should be used for (2)? If it is the personal wallet, what job does the internal wallet have? If it is the internal wallet, why can't we reuse the funds from the personal one for swaps in the other direction?

It appears to me that the ASB only has a single wallet and that is an internal one (internal because we control the keys). The only thing we want is to be able to directly send money out of the ASB without an intermediary transaction through the internal wallet. That is an _optimization_ of existing functionality, not a new, orthogonal feature. However, the current code doesn't make it very clear that we are optimization something. Rather, it suggests that there are reasons why we don't to manage to different sets of keys and I don't see why we should be doing that.

Keep in mind that we sort of have two wallets in the code: We have the ephemeral keys that we use for the protocol and an actual wallet that is used to hold all funds that we receive eventually. s Is that what you are trying to express with "internal" and "personal" wallet?

We don't use the wallet to derive these keys so as expressed above, we don't actually need a wallet instance for Alice at the moment for anything other than getting the two final addresses.

File: swap/src/bin/asb.rs (hunk 11 above)
Comment (thomaseizinger) [reply swap/src/bin/asb.rs:314]: > If a user provides an xpub key he really should not use this xpub otherwise to get addresses.

Why not? It seems perfectly reasonable to wanting to pool together funds into a single wallet from different sources?

> Does BDK support the feature of only scanning from a particular height?

I thought it did but apparently it doesn't. https://docs.rs/bdk/0.7.0/bdk/wallet/struct.Wallet.html#method.sync

The feature of rescan height is actually coupled to the individual blockchain backend in BDK. Since we are using electrum, it doesn't matter.

Comment (bonomat) [reply swap/src/bin/asb.rs:314]: > Why not? It seems perfectly reasonable to wanting to pool together funds into a single wallet from different sources?

It's mostly a privacy concern: Anyone holding your xpub can derive addresses and hence can learn the balances. Meaning, the Asb or Cli can both learn about all your funds and transaction history unrelated to our applications. 

There is also the concern about security of address re-use which we cannot guarantee if addresses are being created outside of our control.

Comment (thomaseizinger) [reply swap/src/bin/asb.rs:314]: Fair. As we don't support any other backends than Electrum for now, it is not a performance concern so we don't need the rescan height actually.

File: swap/src/seed.rs (hunk 3 above)
Comment (bonomat) [reply swap/src/seed.rs:57]: > But at the moment (without this feature) they are, right? Is your claim that the current code and behaviour is wrong?
> [...]
> Keep in mind that we sort of have two wallets in the code: We have the ephemeral keys that we use for the protocol and an actual wallet that is used to hold all funds that we receive eventually. Is that what you are trying to express with "internal" and "personal" wallet?

Wrong might be a strong word but yes, I think the we should have two different wallets if the 'internal' wallet even qualifies to be called as a wallet as it is only used to sign protocol transactions. 

Anyway, changing this is out of scope of this ticket. 

What I wanted to say with: 

> I'm coming to a different conclusion: the code here is wrong.
The two wallets should not be the same.

Is that the from me newly introduced code is wrong. We should take the chance to decouple the two wallets so that there is no relation between them and not derive the xpub from the same xprv as used for the internal wallet. 

This is also useful when the ASB can trade in both directions, i.e. it has one wallet that holds the money and one _wallet_ which is used to sign protocol transactions.

Comment (thomaseizinger) [reply swap/src/seed.rs:57]: > This is also useful when the ASB can trade in both directions, i.e. it has one wallet that holds the money and one wallet which is used to sign protocol transactions.

Okay, I think we finally came to the bottom of all this!

We don't use `bitcoin::Wallet` to sign protocol transactions. Never. All protocol transactions are signed with very specific private keys that are generated completely randomly during execution setup.

The only transactions signed by `bitcoin::Wallet` are:

- Funding `TxLock`, which by itself does not qualify as a protocol transaction as it spends arbitrary inputs
- Withdrawing to an address

All other transactions:

- spending TxLock as redeem
- spending TxLock as cancel
- spending TxCancel as punish
- spending TxCancel as refund

are not signed by `bitcoin::Wallet` but by randomly generated keys.

As such, I don't see why we need **two** `bitcoin::Wallet` instances.

Comment (bonomat) [reply swap/src/seed.rs:57]: riiight... I don't know what I thought and followed blindly the recommendation in the issue...

Since this is the case, we definitely don't need two wallets. In fact, this ticket is way simpler than this PR proposes and can be reduced to very few changes. 

We let the user define a xpub key (or *pub as defined in [SLIP-132](https://github.com/satoshilabs/slips/blob/master/slip-0132.md)) and instantiate the wallet from this key. If it is an xpub key we have a read-only wallet but this is fine if we only swap in one one direction. 
If no key was provided, we derive one from the internal seed as we used to.

Comment (bonomat) [reply swap/src/seed.rs:57]: cc @da-kami :⬆️ this conclusion is crucial if you take over this PR.

**Comment by bonomat** [comment]
This PR was build under a wrong assumption as discovered in this comment: https://github.com/comit-network/xmr-btc-swap/pull/507#discussion_r633943295 

Closing this as it is easier to start from scratch.
